### ISSUE TITLE
Fix `Rack::VERSION` to be more canonical.

### ISSUE THREAD

**Issue by ioquatix** [rack/rack#1843]
`Rack::VERSION` should be something like `"1.2.3"` but it's an `Array` which is pretty atypical. I'd like to fix this to be more consistent with other gems which define it as a `String`.

**Comment by jeremyevans** [comment]
I don't think the benefits of consistency with other gems outweigh the cost of the loss of backwards compatibility.

**Comment by simi** [comment]
What's the policy for backward compatibility for Rack 3?

**Comment by jeremyevans** [comment]
As in an major version change, backwards compatibility changes are allowed.  However, I have a very strong preference for only breaking backwards compatibility when the benefits significantly outweigh the costs.

**Comment by rafaelfranca** [comment]
We could add a new `Rack.version` method that returns a string? That is what we end up doing in Rails.

**Comment by tenderlove** [comment]
I agree with @jeremyevans and @rafaelfranca.  I'm fine if we add a method, but "consistency with other gems" isn't a goal of the project.

**Comment by jeremyevans** [comment]
We already have a `Rack.version` method.  However, it gives the rack protocol version (currently 1.3), not the gem version (which would be 3.0.0).  The gem version is available via `Rack.release`.  I propose we make no changes.

**Comment by ioquatix** [comment]
Does the rack protocol version have any meaning? Can anyone depend on it for anything useful?

**Comment by rafaelfranca** [comment]
Rails also has [`gem_version`](https://github.com/rails/rails/blob/main/railties/lib/rails/gem_version.rb#L6).

I don't have a strong opinion if we should add a method like that or not, but I'm giving us options.

**Comment by ioquatix** [comment]
The concept of compatibility version or "interface version" or "spec version" I don't think has ever been useful.

My gut feeling is we should remove this. We don't have middleware which change their behaviour based on the version of rack nor should we encourage it.

I don't mind if we need a middle state to get to the ideal, but my ideal state is this:

```rack
module Rack
  VERSION = "3.0.0"
end
```

**Comment by ioquatix** [comment]
```
> bundle gem foobar
Creating gem 'foobar'...
MIT License enabled in config
      create  foobar/Gemfile
      create  foobar/lib/foobar.rb
      create  foobar/lib/foobar/version.rb
      create  foobar/foobar.gemspec
      create  foobar/Rakefile
      create  foobar/README.md
      create  foobar/bin/console
      create  foobar/bin/setup
      create  foobar/.gitignore
      create  foobar/.travis.yml
      create  foobar/.rspec
      create  foobar/spec/spec_helper.rb
      create  foobar/spec/foobar_spec.rb
      create  foobar/LICENSE.txt
Initializing git repo in /private/tmp/foobar
Gem 'foobar' was successfully created. For more information on making a RubyGem visit https://bundler.io/guides/creating_gem.html

> cat foobar/lib/foobar/version.rb 
module Foobar
  VERSION = "0.1.0"
end
```

This is literally the recommended default style.

**Comment by jeremyevans** [comment]
If we want to remove/change `Rack::VERSION`/`Rack.version`, we need to deprecate in 3.0 and remove/change in 3.1/4.0.

**Comment by ioquatix** [comment]
I would personally like this gem to be as canonical as possible for several reasons:

- Easier for maintainers to understand.
- Easier for users to understand.
- Less work maintaining bespoke details.
- Better integration with tooling that expects the canonical structure.
- Consistency with other gems.

I would personally like to do this consistently across all `rack/` gems.

I am happy to make a PR according to @jeremyevans's suggestions if no one else has strong objections. I'll also make PRs to other projects to bring them into alignment.

**Comment by ioquatix** [comment]
I went looking for existing usage of `Rack.release` and found one case:

https://github.com/omniauth/omniauth/commit/c85c717f0d2b478ce96729fe40d1694aadebe665

However it's removed since 2020 and was only ever useful for Rack < 1.6.

**Comment by ioquatix** [comment]
Found another example:

https://github.com/MarcGrimme/context-request-middleware/blob/ff6b210711db7de22093ada60a3f988e57f5ba08/lib/context_request_middleware/cookie.rb

**Comment by jeremyevans** [comment]
Closing as this was addressed by #1940
